// @extjs/reactor-webpack-plugin/dist/artifacts.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.buildXML = undefined;
exports.createAppJson = createAppJson;
exports.createJSDOMEnvironment = createJSDOMEnvironment;
exports.createWorkspaceJson = createWorkspaceJson;

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

var _cjson = require('cjson');

var _cjson2 = _interopRequireDefault(_cjson);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var buildXML = exports.buildXML = function buildXML(_ref) {
    var compress = _ref.compress;

    var compression = '';

    if (compress) {
        compression = '\n            then \n            fs \n            minify \n                -yui \n                -from=ext.js \n                -to=ext.js\n        ';
    }

    return ('\n        <project name="simple-build" basedir=".">\n            <!--  internally, watch calls the init target, so need to have one here -->\n            <target name="init"/>\n            <target name="init-cmd">\n                <taskdef resource="com/sencha/ant/antlib.xml"\n                                classpath="${cmd.dir}/sencha.jar"\n                                loaderref="senchaloader"/>\n                <x-extend-classpath>\n                    <jar path="${cmd.dir}/sencha.jar"/>\n                </x-extend-classpath>\n                <x-sencha-init prefix=""/>\n                <x-compile refid="theCompiler"\n                                 dir="${basedir}"\n                                 initOnly="true"\n                                 inheritAll="true">\n                          <![CDATA[\n                          -classpath=${basedir}/manifest.js\n                          load-app\n                              -temp=${basedir}/temp\n                              -tag=App\n                    ]]>\n                  </x-compile>\n            </target>\n            <target name="rebuild">\n               <x-compile refid="theCompiler"\n                          dir="${basedir}"\n                          inheritAll="true">\n                  <![CDATA[\n                  --debug\n                  exclude\n                  -all\n                  and\n                  include\n                  -f=Boot.js\n                  and\n                  concatenate\n                      ext.js\n                  and\n                  exclude\n                  -all\n                  and\n                  # include theme overrides\n                  include\n                    -r\n                    -tag=overrides\n                  and\n                  # include all js files needed for manifest.js\n                  include\n                      -r\n                      -f=manifest.js\n                  and\n                  # exclude the generated manifest file itself,\n                  # since we don\'t want the generated bundle file to create any components\n                  exclude\n                  -f=manifest.js\n                  and\n                  concatenate\n                  +append\n                      ext.js\n                  and\n                  scss\n                      -appName=App\n                      -imageSearchPath=resources\n                      -themeName=triton\n                      -resourceMapBase=.\n                      -output=ext.scss\n                  and\n                  resources\n                      -excludes=-all*.css\n                      -out=resources\n                  and\n                  resources\n                      -model=true\n                      -out=resources\n                  ]]>\n               </x-compile>\n            </target>\n            <target name="build" depends="init-cmd,rebuild">\n               <x-sencha-command dir="${basedir}">\n                   <![CDATA[\n                   fashion\n                       -pwd=.\n                       -split=4095\n                       ' + (compress ? '-compress' : '') + '\n                           ext.scss\n                       ext.css\n                   ' + compression + '\n                   ]]>\n               </x-sencha-command>\n            </target>\n            <target name="watch" depends="init-cmd,build">\n                <x-fashion-watch\n                        refName="fashion-watch"\n                        inputFile="ext.scss"\n                        outputFile="ext.css"\n                        split="4095"\n                        compress="' + (compress ? 'true' : 'false') + '"\n                        configFile="app.json"\n                        fork="true"/>\n                <x-watch compilerRef="theCompiler"\n                         targets="rebuild"/>\n            </target>\n        </project>\n    ').trim();
};

/**
 * Creates the app.json file
 * @param {String} theme The name of the theme to use.
 * @param {String[]} packages The names of packages to include in the build
 */
function createAppJson(_ref2) {
    var theme = _ref2.theme,
        packages = _ref2.packages,
        toolkit = _ref2.toolkit,
        _ref2$overrides = _ref2.overrides,
        overrides = _ref2$overrides === undefined ? [] : _ref2$overrides,
        _ref2$packageDirs = _ref2.packageDirs,
        packageDirs = _ref2$packageDirs === undefined ? [] : _ref2$packageDirs;

    var config = {
        framework: "ext",
        toolkit: toolkit,
        requires: packages,
        overrides: overrides.map(function (dir) {
            return _path2.default.resolve(dir);
        }).concat('jsdom-environment.js'),
        packages: {
            dir: packageDirs.map(function (dir) {
                return _path2.default.resolve(dir);
            })
        },
        output: {
            base: '.',
            resources: {
                path: './resources',
                shared: "./resources"
            }
        }
    };

    // if theme is local add it as an additional package dir
    if (_fs2.default.existsSync(theme)) {
        var packageInfo = _cjson2.default.load(_path2.default.join(theme, 'package.json'));
        config.theme = packageInfo.name;
        config.packages.dir.push(_path2.default.resolve(theme));
    } else {
        config.theme = theme;
    }

    return JSON.stringify(config, null, 4);
}

/**
 * Creates a js file containing code to make Ext JS load properly in jsdom
 * @param {String} targetDir 
 */
function createJSDOMEnvironment(targetDir) {
    return 'window.Ext = Ext;';
}

/**
 * Creates the workspace.json file
 * @param {String} sdk The path to the sdk
 */
function createWorkspaceJson(sdk, packages, output) {
    return JSON.stringify({
        "frameworks": {
            "ext": _path2.default.relative(output, sdk)
        },
        "packages": {
            "dir": ['${workspace.dir}/packages/local', '${workspace.dir}/packages'].concat(packages).join(','),
            "extract": "${workspace.dir}/packages/remote"
        }
    }, null, 4);
}
